File: frameworks/node/hono/Implementation/routes/authorized.ts
import type { MiddlewareHandler } from "hono";
import { HTTPException } from "hono/http-exception";

import type { Routes } from "../app.ts";

// rb:wiring authorized.*
/**
 * The token check, as middleware on the route, which runs before the handler. A token that is not
 * settings.json's is refused by throwing Hono's HTTPException with 403, as Hono's own csrf and
 * ip-restriction middleware refuse a request, and Hono's error handler writes it.
 */
function requireToken(token: string): MiddlewareHandler {
  const accepted = `Bearer ${token}`;
  return async (c, next) => {
    if (c.req.header("authorization") !== accepted) throw new HTTPException(403, { message: "Forbidden" });
    await next();
  };
}
// rb:end

/**
 * authorized: hono/bearer-auth refuses a wrong token with 401 where the corpus asks for 403, and
 * takes no status, so the check is middleware written for the route.
 */
const authorized: Routes = (app, p) => {
  app.get("/authorized/small", requireToken(p.settings.token), (c) => c.json(p.small));
};

export default authorized;

File: frameworks/go/fiber/Implementation/authorized.go
package implementation

import (
	"github.com/gofiber/fiber/v3"
)

// authorizedRoutes put a middleware in front of the handler that refuses any bearer token but
// settings.json's. Fiber's keyauth middleware answers a wrong key with 401, and the corpus asks
// for 403, so the check is a middleware written for the route, returning Fiber's own 403.
func authorizedRoutes(app *fiber.App, p *Payloads) {
	app.Get("/authorized/small", requireToken(p.Settings.Token), func(c fiber.Ctx) error { return c.JSON(&p.Small) })
}

// rb:wiring authorized.*
func requireToken(token string) fiber.Handler {
	expected := "Bearer " + token
	return func(c fiber.Ctx) error {
		if c.Get(fiber.HeaderAuthorization) != expected {
			return fiber.ErrForbidden
		}
		return c.Next()
	}
}
